Source/hydro/trace_ppm.cpp

```diff
@@ -0,0 +1,561 @@
+    qpass_map_p[n] = qpass_map[n];
+  }
+
+  // Trace to left and right edges using upwind PPM
+  AMREX_PARALLEL_FOR_3D(bx, i, j, k,
+  {
+
+    Real rho = q(i,j,k,QRHO);
+
+
+    Real cc = qaux(i,j,k,QC);
+
+#if AMREX_SPACEDIM < 3
+    Real csq = cc*cc;
+#endif
+
+    Real un = q(i,j,k,QUN);
+
+
+    // do the parabolic reconstruction and compute the
+    // integrals under the characteristic waves
+    Real s[5];
+    Real flat = flatn(i,j,k);
+    Real sm;
+    Real sp;
+
+    Real Ip[NQ][3];
+    Real Im[NQ][3];
+
+
+    for (int n = 0; n < NQ; n++) {
+      if (n == QTEMP) continue;
+
+      if (idir == 0) {
+        s[im2] = q(i-2,j,k,n);
+        s[im1] = q(i-1,j,k,n);
+        s[i0]  = q(i,j,k,n);
+        s[ip1] = q(i+1,j,k,n);
+        s[ip2] = q(i+2,j,k,n);
+
+      } else if (idir == 1) {
+        s[im2] = q(i,j-2,k,n);
+        s[im1] = q(i,j-1,k,n);
+        s[i0]  = q(i,j,k,n);
+        s[ip1] = q(i,j+1,k,n);
+        s[ip2] = q(i,j+2,k,n);
+
+      } else {
+        s[im2] = q(i,j,k-2,n);
+        s[im1] = q(i,j,k-1,n);
+        s[i0]  = q(i,j,k,n);
+        s[ip1] = q(i,j,k+1,n);
+        s[ip2] = q(i,j,k+2,n);
+
+      }
+
+      ppm_reconstruct(s, flat, sm, sp);
+      ppm_int_profile(sm, sp, s[i0], un, cc, dtdx, Ip[n], Im[n]);
+
+    }
+
+    // gamma_c
+
+    Real Ip_gc[3];
+    Real Im_gc[3];
+
+    if (idir == 0) {
+        s[im2] = qaux(i-2,j,k,QGAMC);
+        s[im1] = qaux(i-1,j,k,QGAMC);
+        s[i0]  = qaux(i,j,k,QGAMC);
+        s[ip1] = qaux(i+1,j,k,QGAMC);
+        s[ip2] = qaux(i+2,j,k,QGAMC);
+
+    } else if (idir == 1) {
+        s[im2] = qaux(i,j-2,k,QGAMC);
+        s[im1] = qaux(i,j-1,k,QGAMC);
+        s[i0]  = qaux(i,j,k,QGAMC);
+        s[ip1] = qaux(i,j+1,k,QGAMC);
+        s[ip2] = qaux(i,j+2,k,QGAMC);
+
+    } else {
+        s[im2] = qaux(i,j,k-2,QGAMC);
+        s[im1] = qaux(i,j,k-1,QGAMC);
+        s[i0]  = qaux(i,j,k,QGAMC);
+        s[ip1] = qaux(i,j,k+1,QGAMC);
+        s[ip2] = qaux(i,j,k+2,QGAMC);
+
+    }
+
+
+    ppm_reconstruct(s, flat, sm, sp);
+    ppm_int_profile(sm, sp, s[i0], un, cc, dtdx, Ip_gc, Im_gc);
+
+
+    // source terms
+    Real Ip_src[NQSRC][3];
+    Real Im_src[NQSRC][3];
+
+    for (int n = 0; n < NQSRC; n++) {
+
+      // do we even need to trace (non-zero source?)
+#ifndef AMREX_USE_CUDA
+      int do_trace = do_source_trace[n];
+#else
+      int do_trace = 0;
+      if (idir == 0) {
+        for (int b = i-2; b <= i+2; b++) {
+          if (std::abs(srcQ(b,j,k,n)) > 0.0_rt) {
+            do_trace = 1;
+            break;
+          }
+        }
+      } else if (idir == 1) {
+        for (int b = j-2; b <= j+2; b++) {
+          if (std::abs(srcQ(i,b,k,n)) > 0.0_rt) {
+            do_trace = 1;
+            break;
+          }
+        }
+      } else {
+        for (int b = k-2; b <= k+2; b++) {
+          if (std::abs(srcQ(i,j,b,n)) > 0.0_rt) {
+            do_trace = 1;
+            break;
+          }
+        }
+      }
+#endif
+
+      //do_trace = 1;
```

Comment:
Can you remove this debugging statement please?

Reply:
done